prov/mrail/src/mrail_rma.c

```diff
@@ -0,0 +1,428 @@
+/*
+ * Copyright (c) 2018 Intel Corporation, Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "mrail.h"
+
+static inline void mrail_iov_to_rail(const struct iovec *iov, const void *desc,
+		const uint32_t rail, struct iovec *out_iov, void **out_desc)
+{
+	const struct mrail_mr *mrail_mr = desc;
+
+	//TODO: add base address from mrail_mr
+	out_iov->iov_base 	= (void*)((uintptr_t)iov->iov_base);
+	out_iov->iov_len	= iov->iov_len;
+
+	*out_desc = (mrail_mr ? fi_mr_desc(mrail_mr->rails[rail].mr) : NULL);
+}
+
+static inline void mrail_rma_iov_to_rail(const struct fi_rma_iov *rma_iov,
+		const uint32_t rail, struct fi_rma_iov *out_rma_iov)
+{
+	struct mrail_addr_key *mr_map = (struct mrail_addr_key *)rma_iov->key;
+
+	//TODO: add base address from mrail_addr_key
+	out_rma_iov->addr 	= rma_iov->addr;
+	out_rma_iov->len	= rma_iov->len;
+	out_rma_iov->key	= mr_map[rail].key;
+}
+
+static void mrail_subreq_to_rail(struct mrail_subreq *subreq, uint32_t rail,
+		struct iovec *out_iovs, void **out_descs,
+		struct fi_rma_iov *out_rma_iovs)
+{
+	size_t i;
+
+	for (i = 0; i < subreq->iov_count; ++i) {
+		mrail_iov_to_rail(&subreq->iov[i], subreq->descs[i], rail,
+				&out_iovs[i], &out_descs[i]);
+	}
+
+	for (i = 0; i < subreq->rma_iov_count; ++i) {
+		mrail_rma_iov_to_rail(subreq->rma_iov, rail, &out_rma_iovs[i]);
+	}
+}
+
+static ssize_t mrail_try_posting_subreq(uint32_t rail,
```

Comment:
I think it's okay to replace 'try_posting' with just 'post', with the assumption that the post may fail.